colbuilder: missing cast between integers of different widths

roachtest.sqlsmith/setup=rand-tables/setting=no-ddl [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3067408&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3067408&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-ddl) on master @ [ecf3f618dc42b025f542aeec89bb0630a3f36fcd](https://github.com/cockroachdb/cockroach/commits/ecf3f618dc42b025f542aeec89bb0630a3f36fcd):


```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-ddl/run_1
	sqlsmith.go:215,sqlsmith.go:251,test_runner.go:755: error: pq: internal error: unexpected error from the vectorized engine: interface conversion: coldata.Column is coldata.Int64s, not coldata.Int16s
		stmt:
		WITH
			with_2973 (col_17951) AS (SELECT * FROM (VALUES (ARRAY[784675529:::OID,958832258:::OID])) AS tab_8864 (col_17951))
		SELECT
			cte_ref_884.col_17951 AS col_17952,
			st_zmflag('010200004003000000DC2284F164A3FC41806B05B911B8E6C1C852490C60ADEA415AF5A2B2202A0242B7BA23FF30D300C230E9F6CDFDD6FB411EF117E00289F04150024BA6AB77FB41B8F6FAC2023AD241':::GEOMETRY::GEOMETRY)::INT2
				AS col_17953,
			tab_8865.col1_4 AS col_17954,
			e'O4\x03lK\x7f#\x12+':::STRING AS col_17955,
			tab_8865.col1_11 AS col_17956,
			'01060000800600000001030000800100000004000000E0774FFCBB32014248ECA130526BEBC170ACE25209CAC74134E73EE52C45D9C18C7DEA7B6CCDFF4190ED234D2524E141632B7169DE0A00C2B8B4BEA5001AFB41B6F6FE6C768EEFC1E0774FFCBB32014248ECA130526BEBC170ACE25209CAC7410103000080010000000C0000007D4E3F4AE94DF6C1CC082A5B1A16E3415E7808EE1CD200C2A38698B4018EF0C12832B2D8772CE24178BD7DE5C48CEA415AC0B1044ED7FFC134F52193A3E2E0C1244415085D22F4C1C01BAF0FC2E7BE410227E487C3ADE2C10062DB8F22D886413C3C4A471DBFFB41162766055C3BFAC14FD61CC39014F7C1700AEF3D98F90042C4200F4DE8AAE241466B13063F7BF341FC25D75EB696FA41A858E00144A6ED411128DAD1696CF9C1E8DCEDB42D8AFC416C302D986D20F74102A75AC42774ECC1D09762FD67F5C6410A8FA571D3BCF7416A3A534F8D91EDC13C9B3BCB8945D8C1E6E93F745F600142B02D97319D4FDC41A35946784709F7C1A4809371454AFD41C8204A1447BEF6417D4E3F4AE94DF6C1CC082A5B1A16E3415E7808EE1CD200C20103000080010000000A000000B05480C3D9BBEEC1A01F079E4CE4CBC18842ABEEDF57FD414486C8F9099DEEC1A0B21FA7DAD3F1C144A2BB58873AEE41D098981D0FE5DDC1F74368046F43FAC1FF5B92A3DFB9FBC1B457C6F2935B004264CC8D4D375DFCC1D6FA8BAA2143E5C11016C97A6D88C04160FF11897E4DB241204EF8854251BF41EE256AADBAA9F7415662FDE3AB950142E8171082ED42DE41AAFED22777F0E4C16A21B97126380142A28D28EEEA7BEAC167DBD514582900C2704AB2BD5742FB413BF87E6F990902C2077971989CF8F8C1785433C80B24DF416AC9C6D9691CFAC1B05480C3D9BBEEC1A01F079E4CE4CBC18842ABEEDF57FD410103000080010000000A000000B480635D50C5EBC1E6F4FD088A5EF9C1C06AEFC62E7BA7C1D8A7D21F4B56D4C116224C9B1FB3EAC130EACA34DB68E5416095E3CDEBE1CC4109CB56BD974801C214F81D9B3BBAF441128218F38DE7F4419364D3E9540FF4C1C88F00FAEFF3F2C108DE4F61D359D0413801484FAB5FFB41A2257E3870F900C2F45C0BD8889FD8C100DE13018567D7419C626CFD8C83DAC123D4DE70FA0FF4C16870F1473455D34118B2D8401BE1DB4152CEC900D657FFC1B003D8675F40CA4180DAE3ADEF00C841A8FB95B0D2A5E7C1D8F82995F85CD3C11462182DFC8BFFC1B480635D50C5EBC1E6F4FD088A5EF9C1C06AEFC62E7BA7C10103000080010000000500000080429F780908C2C1F583A4E3EF06FFC124FE9339B7F9ECC10020ECFF1912F941E1D189B4F28FF9C1E0FB44DF32CAEC4120E9639507C2BC4120B932279DCED94140D09FE13A6CE3410AD8F2DA0900F8C1A8B32B17C0B0EC41EAD4174B1B09F54180429F780908C2C1F583A4E3EF06FFC124FE9339B7F9ECC101030000800100000005000000300D2D97B509C3C16E4D7DDF85FD00C254CECC89FD35EB417C6D46644DFEE04147F43A772328F1C12036DDB3E00F024230CE509CDD87E7C108ACE462F683D841127385E245C2E2C1C4D80F38A815EFC1AA06F65BBB9CE6C14E523C57F07000C2300D2D97B509C3C16E4D7DDF85FD00C254CECC89FD35EB41':::GEOMETRY
				AS col_17957
		FROM
			defaultdb.public.table1@[0] AS tab_8865, with_2973 AS cte_ref_884
		LIMIT
			10:::INT8;
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
# From https://go.crdb.dev/p/roachstress, perhaps edited lightly.
caffeinate ./roachstress.sh sqlsmith/setup=rand-tables/setting=no-ddl
```
</p>
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #66233 roachtest: sqlsmith/setup=rand-tables/setting=no-ddl failed [C-test-failure O-roachtest O-robot branch-release-21.1]
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*sqlsmith/setup=rand-tables/setting=no-ddl.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -532,12 +532,8 @@ func takeOverMetaInfo(target *colexecargs.OpWithMetaInfo, inputs []colexecargs.O
 // createAndWrapRowSource takes a processor spec, creating the row source and
 // wrapping it using wrapRowSources. Note that the post process spec is included
 // in the processor creation, so make sure to clear it if it will be inspected
-// again. NewColOperatorResult is updated with the new OutputTypes and the
-// resulting Columnarizer if there is no error. The result is also annotated as
-// streaming because the resulting operator is not a buffering operator (even if
-// it is a buffering processor). This is not a problem for memory accounting
-// because each processor does that on its own, so the used memory will be
-// accounted for.
+// again. opResult is updated with the new ColumnTypes and the resulting
+// Columnarizer if there is no error.
 // - causeToWrap is an error that prompted us to wrap a processor core into the
 // vectorized plan (for example, it could be an unsupported processor core, an
 // unsupported function, etc).
@@ -1520,14 +1516,11 @@ func (r opResult) wrapPostProcessSpec(
 	}
 	inputToMaterializer := colexecargs.OpWithMetaInfo{Root: r.Root}
 	takeOverMetaInfo(&inputToMaterializer, args.Inputs)
-	if err := r.createAndWrapRowSource(
+	// createAndWrapRowSource updates r.ColumnTypes accordingly.
+	return r.createAndWrapRowSource(
 		ctx, flowCtx, args, []colexecargs.OpWithMetaInfo{inputToMaterializer},
 		[][]*types.T{r.ColumnTypes}, noopSpec, factory, causeToWrap,
-	); err != nil {
-		return err
-	}
-	r.ColumnTypes = resultTypes
-	return nil
+	)
 }
 
 // planPostProcessSpec plans the post processing stage specified in post on top

--- a/pkg/sql/logictest/testdata/logic_test/dist_vectorize
+++ b/pkg/sql/logictest/testdata/logic_test/dist_vectorize
@@ -484,3 +484,30 @@ SELECT CASE WHEN a>1 THEN b*2 ELSE b*10 END FROM t1
 ----
 NULL
 NULL
+
+# Regression test for the wrapped row-execution processor not satisfying the
+# width of an integer column during a cast and the vectorized engine not
+# performing the cast to the integer of the desired width (#66306).
+statement ok
+CREATE TABLE t66306 (k DECIMAL PRIMARY KEY);
+INSERT INTO t66306 VALUES (1);
+ALTER TABLE t66306 EXPERIMENTAL_RELOCATE VALUES (ARRAY[2], 1);
+
+query II
+SELECT 1::INT2, k::INT8 FROM t66306;
+----
+1 1
+
+# Sanity check that the wrapped processor is planned for the query above. If it
+# no longer is, we should adjust the query here and above.
+query T
+EXPLAIN (VEC) SELECT 1::INT2, k::INT8 FROM t66306;
+----
+│
+├ Node 1
+│ └ *colrpc.Inbox
+└ Node 2
+  └ *colrpc.Outbox
+    └ *colexecbase.castInt64Int16Op
+      └ *rowexec.noopProcessor
+        └ *colfetcher.ColBatchScan

